mod-source-record-storage-server/src/main/java/org/folio/services/util/AdditionalFieldsUtil.java

```diff
@@ -0,0 +1,92 @@
+import java.io.ByteArrayOutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+
+/**
+ * Util to work with additional fields
+ */
+public class AdditionalFieldsUtil {
+
+  public static final String TAG_999 = "999";
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(AdditionalFieldsUtil.class);
+  private static final char INDICATOR = 'f';
+
+  private AdditionalFieldsUtil() {
+  }
+
+  /**
+   * Adds field if it does not exist and a subfield with a value to that field
+   *
+   * @param record   record that needs to be updated
+   * @param field    field that should contain new subfield
+   * @param subfield new subfield to add
+   * @param value    value of the subfield to add
+   * @return true if succeeded, false otherwise
+   */
+  public static boolean addFieldToMarcRecord(Record record, String field, char subfield, String value) {
+    boolean result = false;
+    try (ByteArrayOutputStream os = new ByteArrayOutputStream()) {
+      if (record != null && record.getParsedRecord() != null && record.getParsedRecord().getContent() != null) {
+        MarcReader reader = buildMarcReader(record);
+        MarcWriter streamWriter = new MarcStreamWriter(new ByteArrayOutputStream());
+        MarcJsonWriter jsonWriter = new MarcJsonWriter(os);
+        MarcFactory factory = MarcFactory.newInstance();
+        if (reader.hasNext()) {
+          org.marc4j.marc.Record marcRecord = reader.next();
+          VariableField variableField = getSingleFieldByIndicators(marcRecord.getVariableFields(field), INDICATOR, INDICATOR);
+          DataField dataField;
+          if (variableField != null
+            && ((DataField) variableField).getIndicator1() == INDICATOR
+            && ((DataField) variableField).getIndicator2() == INDICATOR
+          ) {
+            dataField = (DataField) variableField;
+            marcRecord.removeVariableField(variableField);
+          } else {
+            dataField = factory.newDataField(field, INDICATOR, INDICATOR);
+          }
+          dataField.addSubfield(factory.newSubfield(subfield, value));
+          marcRecord.addVariableField(dataField);
+          // use stream writer to recalculate leader
+          streamWriter.write(marcRecord);
+          jsonWriter.write(marcRecord);
+          record.setParsedRecord(record.getParsedRecord().withContent(new JsonObject(new String(os.toByteArray())).encode()));
+          result = true;
+        }
+      }
+    } catch (Exception e) {
+      LOGGER.error("Failed to add additional subfield {} for field {} to record {}", e, subfield, field, record.getId());
+    }
+    return result;
+  }
+
+  private static MarcReader buildMarcReader(Record record) {
+    return new MarcJsonReader(new ByteArrayInputStream(record.getParsedRecord().getContent().toString().getBytes(StandardCharsets.UTF_8)));
+  }
+
+  private static VariableField getSingleFieldByIndicators(List<VariableField> list, char ind1, char ind2) {
```

Comment:
@RuslanLavrov why did you pass there these arguments: "ind1", "ind2"? Regarding current logic, they are always  "f".